lib/helpers.ts

```diff
@@ -19,39 +20,53 @@ export function combineText(post: any) {
   return text;
 }
 
-export async function processPost(
-  post: any,
-  teamId: string,
-  keywords: string[] = []
-): Promise<{ status: "present" | "absent" | "deleted" | "error" }> {
-  /* Process post to determine if it contains our keywords */
-  if (post.deleted) {
-    return { status: "deleted" };
-  }
-  const text = combineText(post);
-
-  for (let i = 0; i < keywords.length; i++) {
-    const keyword = keywords[i];
-
-    /* 
-      This regex matches all instances of a keyword that are not preceded or proceeded by alphabets
-      e.g. keyword = `vite`:
-      - ...the Vite framework is... => matches
-      - ...we use vite. It is pretty... => matches
-      - ...my app is at https://vite.vercel.app/ => matches
-      - ...please send me an invite for... => does not match
-    */
-    if (RegExp(`(?<![A-Za-z])${keyword}(?![A-Za-z]+)`, "gmi").test(text)) {
-      try {
-        await sendSlackMessage(post.id, teamId); // send message to Slack
-        return { status: "present" };
-      } catch (err) {
-        console.log(err);
-        return { status: "error" };
+export function postScanner(teamsAndKeywords: TeamAndKeywords[]) {
+  const keywordMapping = new Map() as Map<string, string[]>;
+  const keywords = new Set();
+
+  for (const team of teamsAndKeywords) {
+    for (const keyword of team.keywords) {
+      keywords.add(keyword);
+
+      let teams = keywordMapping.get(keyword);
+      if (teams === undefined) {
+        teams = [];
+        keywordMapping.set(keyword, teams);
       }
+      teams.push(team.teamId);
     }
   }
-  return { status: "absent" };
+
+  const keywordArray = Array.from(keywords) as string[];
+  const boundaries = keywordArray.map(
+    (keyword) => `\\b${regexEscape(keyword)}\\b`
+  );
+
+  const scanner = new RegExp(`(${boundaries.join(")|(")})`, "gi"); // create a regex that matches all keywords
+
+  return (post: any): Set<string> => {
+    const text = combineText(post); // combine text from post's title, text, and url
+    const teamsInterestedInThisPost = new Set() as Set<string>; // set of team IDs that are interested in this post
+
+    text.replace(scanner, (_, ...terms) => {
```

Comment:
Alternatively, we can use:

```js
let match;
while ((match = scanner.exec(text))) {
  for (let i = 0; i < keywordArray.length; i++) {
    if (match[i + 1] !== undefined) {
      //...
    }
  }
}

Reply:
are there any performance trade offs?

Reply:
I've thought about using `.test()` or `.exec()` as well

Reply:
Gonna merge for now! :)

Reply:
It'll avoid the string concat performance issues in `replace`. `test` won't work, because it'll only return a boolean, but `exec` should be fast.